### PR TITLE
Firefox peer code review PR

### PR DESCRIPTION
**Pull request by motin**
This PR is not for merging, but to be able to review the changes since v1 of this study.

Between the v1 code (at https://github.com/florian/federated-learning-addon) and the beginning of the commits in this PR, I merged in the template and retrofitted the v1 logic into the template. This makes it easier to review code changes since the template-imposed structure has been kept intact through the development of v2.

### CODE REVIEW

File: docs/DEV.md [hunk 1]
```diff
@@ -179,32 +198,86 @@ ln -s Firefox\ Nightly.app FirefoxNightly.app
...
-
-### Core components
-
-* `lib/awesomeBarObserver.js`: Everything related to observing awesome bar interactions, using the events emitted by the `awesomeBar` Experiment API.
-* `lib/synchronization.js`: Everything related to the federated learning protocol. Currently that means sending weight updates back using Telemetry and reading the current model from S3
-* `lib/optimization.js`: For computing model updates
-* `studySetup.js` is adapted from [`shield-studies-addon-utils`](https://github.com/mozilla/shield-studies-addon-utils) and configures the study
-* `feature.js` connects everything.
-
 ## Directory Structure and Files
```
Comment (rhelmer) [inline docs/DEV.md:209]: I find keeping this sort of thing up to date is difficult, if you think it's worth the effort in this case that's OK but something to consider.

File: src/feature.js [hunk 2]
```diff
@@ -1,69 +1,246 @@
...
    */
-  async cleanup() {}
-
+  async cleanup() {
+    if (this.awesomeBarObserver) {
+      await this.awesomeBarObserver.stop();
+    }
+    await browser.study.logger.log("Cleaning up study-specific prefs");
+    const promises = [];
+    for (let i = 0; i < FRECENCY_PREFS.length; i++) {
```
Comment (rhelmer) [inline src/feature.js:236]: You could use `for...of` here to avoid using `length` here and `[i]` below if you want

**Review by rhelmer — APPROVED** [review]
I can't find anything actually wrong, just a few nits. Nice work!

File: src/lib/awesomeBarObserver.js [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+      {
+        observedEventsSinceLastFocus: this.observedEventsSinceLastFocus,
+      },
+    ]);
+
+    try {
+      const focusEvent = this.observedEventsSinceLastFocus
+        .slice()
+        .reverse()
+        .find(observedEvent => {
```
Comment (rhelmer) [inline src/lib/awesomeBarObserver.js:57]: Oh also not super important but this seems like a common pattern, I wonder if you want a function.

File: src/lib/awesomeBarObserver.js [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+    ];
+    const styles = new Set(styleString.split(/\s+/));
+    const isNonBookmarkOrHistoryStyle = NON_BOOKMARK_OR_HISTORY_STYLES.some(s =>
+      styles.has(s),
+    );
+    return !isNonBookmarkOrHistoryStyle;
+  }
+
+  /**
+   * @param {string} searchString Self-explanatory hopefully
```
Comment (rhelmer) [inline src/lib/awesomeBarObserver.js:448]: The "Self-explanatory hopefully" comment seems redundant - I'd either let `searchString` stand on its own then.

File: docs/TELEMETRY.md [hunk 5]
```diff
@@ -31,133 +35,223 @@ This information is used to evaluate the quality of the model.
...
-      -0.8,
-      0.29,
-      0.42,
-      0.89,
-      0.39,
-      0.54,
-      0.78
-    ]
+  "model_version": 140,
+  "frecency_scores": [38223, 3933.4, 304933.3, 21],
```
Comment (crankycoder) [inline docs/TELEMETRY.md:87]: The `frecency_scores` numbers should be integers.

Comment (crankycoder) [reply docs/TELEMETRY.md:87]: oops. nevermind - the schema in v1 was floats.

File: docs/TELEMETRY.md [hunk 6]
```diff
@@ -31,133 +35,223 @@ This information is used to evaluate the quality of the model.
...
-| `num_chars_typed`           | integer           | how many characters did the user type? (should be minimized by the optimization process)            |
-| `frecency_scores`           | array of integers | what scores did the model assign to the suggestions?                                                |
+| name                                             | type              | description                                                                                                                                |
+| ------------------------------------------------ | ----------------- | ------------------------------------------------------------------------------------------------------------------------------------------ |
+| `model_version`                                  | integer           | the version of the model that all the other data is based on                                                                               |
+| `frecency_scores`                                | array of integers | what scores did the model assign to the suggestions?                                                                                       |
```
Comment (crankycoder) [inline docs/TELEMETRY.md]: @motin - this array should be floats as per https://github.com/mozilla-services/mozilla-pipeline-schemas/pull/277/files#diff-5c2b8d38789e1c1028cafdb59c201e47R9

Comment (motin) [reply docs/TELEMETRY.md]: Good catch!

File: docs/DEV.md (hunk 1 above)
Comment (motin) [reply docs/DEV.md:209]: Me too. I just update it once per major version basically. Adding a note to the template that this section is extraordinary optional.

File: src/lib/awesomeBarObserver.js (hunk 4 above)
Comment (motin) [reply src/lib/awesomeBarObserver.js:448]: Me too. I have a note on updating the eslint configuration to not complain in these cases

File: src/lib/awesomeBarObserver.js (hunk 3 above)
Comment (motin) [reply src/lib/awesomeBarObserver.js:57]: True, it did end up being a pattern :) Will wrap in a function in future revisions

**Pull request closed without merging**